cloudbridge/cloud/providers/gce/services.py

```diff
@@ -872,3 +1024,80 @@ def create(self, name, volume, description=None):
             return snapshots[0]
         else:
             return None
+
+
+class GCSObjectStoreService(BaseObjectStoreService):
+
+    def __init__(self, provider):
+        super(GCSObjectStoreService, self).__init__(provider)
+
+    def get(self, bucket_id):
+        """
+        Returns a bucket given its ID. Returns ``None`` if the bucket
+        does not exist or if the user does not have permission to access the
+        bucket.
+        """
+        try:
+            response = (self.provider
+                            .gcp_storage
+                            .buckets()
+                            .get(bucket=bucket_id)
+                            .execute())
+            if 'error' in response:
+                # response['error']['code'] is 404 if the bucket does not exist
+                # and 403 if the user does not have permission to access it.
+                if response['error']['code'] not in (403, 404):
+                    cb.log.warning('Unexpected error code (%d) when accessing '
+                                   'bucket %s', response['error']['code'],
+                                   bucket_id)
+                return None
+            return GCSBucket(self.provider, response)
+        except:
+            return None
+
+    def find(self, name, limit=None, marker=None):
+        """
+        Searches in bucket names for a substring.
+        """
+        buckets = [bucket for bucket in self.list() if name in bucket.name]
+        return ClientPagedResultList(self.provider, buckets, limit=limit,
+                                     marker=marker)
+
+    def list(self, limit=None, marker=None):
+        """
+        List all containers.
+        """
+        try:
+            response = (self.provider
+                            .gcp_storage
+                            .buckets()
+                            .list(project=self.provider.project_name)
+                            .execute())
+            if 'error' in response or 'items' not in response:
+                return []
+            buckets = [GCSBucket(self.provider, bucket)
+                       for bucket in response['items']]
+            return ClientPagedResultList(self.provider, buckets,
```

Comment:
@chiniforooshan I think it would be much better to use a ServerPagedResultList so that filtering happens more efficiently. I believe the bucket.list() method supports maxResults and pageToken which would make this possible.